### PR TITLE
fixes android-ndk regression with CMakeToolchain

### PR DESCRIPTION
**Pull request by AlexRamallo**
Specify library name and version:  **android-ndk/r25**

A recent change of this package added the cmake toolchain file from the NDK as a user toolchain, which is problematic because the CMakeToolchain generator also adds this, resulting in breakage for packages that use this generator, as described in issue #12723 

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
```
A recent change of this package added the cmake toolchain file from the NDK as a user toolchain, which is problematic because the CMakeToolchain generator also adds this, resulting in breakage for packages that use this generator, as described in issue https://github.com/conan-io/conan-center-index/issues/12723
```

Sure android toolchain is injected, like it was in conan v1 through `CONAN_CMAKE_TOOLCHAIN_FILE`. This issue you link is likely another issue, and if it works without the toolchain it means that there is something else wrong in this recipe (or maybe the android toolchain is broken but it would be quite annoying).

**Comment by AlexRamallo** [comment]
As mentioned in the linked issue, the problem is that the toolchain is included twice, [here](https://github.com/conan-io/conan/blob/6fc624d4c1ec165618f9426172b2b0ec569b4de4/conan/tools/cmake/toolchain/blocks.py#L293) and [here](https://github.com/conan-io/conan/blob/6fc624d4c1ec165618f9426172b2b0ec569b4de4/conan/tools/cmake/toolchain/blocks.py#L577)

The build fails due to the UserToolchain block being included first, before the important variables `ANDROID_PLATFORM` and `ANDROID_ABI` are set. The reported error message:

> clang: error: the clang compiler does not support '-march=armv7-a'

Happens because the android toolchain file does not choose the correct flags due to those missing variables.

I don't know if the proper solution is to change CMakeToolchain to deal with this edge case, but I do know that, at the moment, any package that relies on CMakeToolchain is broken when targeting Android, and this PR provides at least a temporary fix until a proper one can be implemented. (at least on Conan v1. I haven't tested v2)

**Comment by SpaceIm** [comment]
> As mentioned in the linked issue, the problem is that the toolchain is included twice, [here](https://github.com/conan-io/conan/blob/6fc624d4c1ec165618f9426172b2b0ec569b4de4/conan/tools/cmake/toolchain/blocks.py#L293) and [here](https://github.com/conan-io/conan/blob/6fc624d4c1ec165618f9426172b2b0ec569b4de4/conan/tools/cmake/toolchain/blocks.py#L577)

My gosh.

@memsharded @lasote Isn't conan trying to be too smart here?

**Comment by lasote** [comment]
Hi, I'm still trying to guess what is needed to be fixed. This PR removes the `self.conf_info.prepend("tools.cmake.cmaketoolchain:user_toolchain", toolchain)` but keeps the `self.conf_info.define("tools.android:ndk_path", self.package_folder)` that is the correct one for this case. 

So, what is the problem with this PR? Is it breaking 1.X builds or...?

**Comment by SpaceIm** [comment]
> Hi, I'm still trying to guess what is needed to be fixed. This PR removes the `self.conf_info.prepend("tools.cmake.cmaketoolchain:user_toolchain", toolchain)` but keeps the `self.conf_info.define("tools.android:ndk_path", self.package_folder)` that is the correct one for this case.


How could a `android-ndk` recipe maintainer know that, since other recipes like `emsdk` have to inject toolchain file themself. In one case it's handled by conan under the hood if `tools.android:ndk_path` is defined, is it documented?

> So, what is the problem with this PR? Is it breaking 1.X builds or...?

No, no problem for conan v1 recipes, android-ndk recipe still have to define `CONAN_CMAKE_TOOLCHAIN_FILE`.

**Comment by lasote** [comment]
> How could a android-ndk recipe maintainer know that, since other recipes like emsdk have to inject toolchain file themself. For android-ndk, there is an hardcoded logic in conan (in CMakeToolchain generator) under the hood if tools.android:ndk_path is defined, is it documented?

Only the conf entires at https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html

> In the long term, do you plan to hardcode more things for other SDK, so that it could break existing sdk recipes?

I would love to have the answer. In the case of Android NDK the conf was pretty useful because it is the natural flow when you download the NDK and you want to indicate where is it. I assume if this happen again with another ecosystem we could study the introduction of a new conf.

File: recipes/android-ndk/all/conanfile.py
Comment (SpaceIm) [inline recipes/android-ndk/all/conanfile.py:289]: maybe add a comment explaining why we must not set `tools.cmake.cmaketoolchain:user_toolchain` in this recipe.

**Review by SpaceIm — APPROVED** [review]

**Comment by prince-chrismc** [comment]
👋 @lasote when you have a change too follow up this it outside of my skill set

**Review by lasote — APPROVED** [review]

**Comment by bog-dan-ro** [comment]
I tested this patch and I can confirm it fixes the android build problems.
Is it possible to merge this fix ASAP? Without it I can't use conan for android anymore ...
Also is it possible to add Android checks to your CI, this way in the future this kind of problems are prevented.

**Comment by uilianries** [comment]
@bog-dan-ro It will be merged automatically, after passing by all tests.

**Comment by AlexRamallo** [comment]
It seems like it's taking a really long time to get me access for some reason. Not sure what the hold up is, but this issue affects a very large number of packages and makes CCI unusable for anyone targeting Android.

I think it would be better if someone who already has access (like @SpaceIm) creates a new PR with the fix so that this gets merged quicker, especially since it's just one line of code being changed.

If someone does that, I'll close this issue.

**Review by jwillikers — APPROVED** [review]

**Comment by uilianries** [comment]
@AlexRamallo It's taking much longer than usual. We have PRs related to Boost and Qt, both consumes several hours and delay other jobs due limited resources.

**Comment by SpaceIm** [comment]
@uilianries it's not hours but days here. The last commit in this PR has been pushed 1 week ago.

**Comment by uilianries** [comment]
@SpaceIm Yes, we noted that huge delay, but the CI is not stuck, it's working:

Screenshot 2022-09-12 at 17-57-23 Branches (13) prod » cci Jenkins

But it's more much busy. Boost takes almost an entire day to build all versions and now, with 1.80.0, will take more time. Also, out biggest bottleneck is Macos so far.

**Comment by SpaceIm** [comment]
https://github.com/conan-io/conan-center-index/pull/12807 shouldn't take so long in c3i. It doesn't modify RREV of previous versions.
So it should be a full build of 1.80.0 only, and for previous versions only a download of existing binaries in conan-center to run test_package.

**Pull request merged**
